File: src/main/java/net/maple/handlers/user/UserHitHandler.java
package net.maple.handlers.user;

import client.Character;
import client.Client;
import net.maple.handlers.PacketHandler;
import net.maple.packets.CharacterPackets;
import util.packet.PacketReader;

public class UserHitHandler extends PacketHandler {

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        Character chr = c.getCharacter();

        int timestamp = reader.readInteger();
        byte type = reader.readByte();
        byte guard = reader.readByte();
        int dmg = reader.readInteger();

        switch (type) {
            case DamageType.MOB_PYSHICAL:
            case DamageType.MOB_MAGIC:
                int mobId = reader.readInteger();
                int objId = reader.readInteger();
                byte left = reader.readByte();
                byte top = reader.readByte();
                byte relativeDir = reader.readByte();
                byte damageMissed = reader.readByte();
                byte v284x = reader.readByte();
                CharacterPackets.showDamage(chr, type, dmg, mobId, left);
                chr.modifyHealth(-dmg);
                break;
            case DamageType.OBSTACLE:
                CharacterPackets.showDamage(chr, type, dmg, 0, (byte) 0);
                chr.modifyHealth(-dmg);
                break;
            default:
                System.err.println("[UserHitHandler] Unknown damage type (" + type + ")");
                break;
        }
    }

    private static final class DamageType {

        // these nexon enums seem to be wrong
        //public static final byte MOB_PYSHICAL = 0x0;
        //public static final byte MOB_MAGIC = 0xFFFFFFFF;
        public static final byte MOB_MAGIC = 0x0;
        public static final byte MOB_PYSHICAL = 0xFFFFFFFF;
        public static final byte COUNTER = 0xFFFFFFFE;
        public static final byte OBSTACLE = 0xFFFFFFFD;
        public static final byte STAT = 0xFFFFFFFC;

        private DamageType() {
        }
    }
}

File: src/main/java/net/maple/handlers/user/UserHitHandler.kt
package net.maple.handlers.user

import client.Client
import net.maple.handlers.PacketHandler
import net.maple.packets.CharacterPackets.showDamage
import util.logging.LogType
import util.logging.Logger.log
import util.packet.PacketReader

class UserHitHandler : PacketHandler {

    override fun handlePacket(reader: PacketReader, c: Client) {
        val chr = c.character

        val timestamp = reader.readInteger()
        val type = reader.readByte()
        val magicElemAttr = reader.readByte() // Element - 0x00 = elementless, 0x01 = ice, 0x02 = fire, 0x03 = lightning
        val dmg = reader.readInteger()

        when (type) {
            DamageType.MOB_PHYSICAL, DamageType.MOB_MAGIC -> {
                val mobId = reader.readInteger()
                val objId = reader.readInteger()
                val left = reader.readByte()
                val top = reader.readByte()
                val relativeDir = reader.readByte()
                val damageMissed = reader.readByte()
                val v284x = reader.readByte()

                chr.showDamage(type, dmg, mobId, left)
            }
            DamageType.OBSTACLE -> {
                chr.showDamage(type, dmg, 0, 0.toByte())
            }
            else -> {
                log(LogType.UNCODED, "Unknown damage type ($type)", this, c)
                return
            }
        }
        chr.modifyHealth(-dmg)
    }

    private object DamageType {
        // these nexon enums seem to be wrong
        //public static final byte MOB_PYSHICAL = 0x0;
        //public static final byte MOB_MAGIC = 0xFFFFFFFF;
        const val MOB_MAGIC: Byte = 0x00
        const val MOB_PHYSICAL: Byte = -0x01
        const val COUNTER: Byte = -0x02
        const val OBSTACLE: Byte = -0x03
        const val STAT: Byte = -0x04
    }
}